mart/attack/perturber.py

```diff
@@ -51,41 +55,103 @@ def __init__(
         """
         super().__init__()
 
-        self.initializer = initializer
+        # Modality-neutral objects.
         self.optimizer_fn = optimizer
-        self.composer = composer
-        self.gradient_modifier = gradient_modifier or GradientModifier()
-        self.projector = projector or Projector()
         self.gain_fn = gain
         self.objective_fn = objective
 
+        # Replace None with nop().
+        gradient_modifier = gradient_modifier or GradientModifier()
+        projector = projector or Projector()
+
+        # Modality-specific objects.
+        # Backward compatibility, in case modality is unknown, and not given in input.
+        if not isinstance(initializer, dict):
+            initializer = {"default": initializer}
+        if not isinstance(gradient_modifier, dict):
+            gradient_modifier = {"default": gradient_modifier}
+        if not isinstance(projector, dict):
+            projector = {"default": projector}
+        if not isinstance(composer, dict):
+            composer = {"default": composer}
+
+        # Backward compatibility, in case optimization parameters are not given.
+        if optim_params is None:
+            optim_params = {modality: {} for modality in initializer.keys()}
+
+        # Modality-specific objects.
+        self.initializer = initializer
+        self.gradient_modifier = gradient_modifier
+        self.projector = projector
+        self.composer = composer
+        self.optim_params = optim_params
+
         self.perturbation = None
 
-    def configure_perturbation(self, input: torch.Tensor | tuple):
-        def create_and_initialize(inp):
-            pert = torch.empty_like(inp, dtype=torch.float, requires_grad=True)
-            self.initializer(pert)
+    def configure_perturbation(self, input: torch.Tensor | tuple | tuple[dict[str, torch.Tensor]]):
+        def create_and_initialize(data, *, input, target, modality="default"):
```

Comment:
Can you make `"default"` a constant somewhere?

Reply:
I made it a class constant.